# Regular and Multi-class One-vs-all Perceptron & Logistic Regression
## Cyril Gorlla
## University of California, San Diego

We look at posts made in six different internet newsgroups: comp.windows.x, rec.sport.baseball, sci.med, misc.forsale, talk.politics.mideast and talk.religion.misc – that
correspond to labels $[1,...,6]$ respectively. The posts have been converted to feature vectors, with each feature signifying the count of a word.

*Source*: [20 Newsgroups Dataset](http://qwone.com/~jason/20Newsgroups/), originally collected by Ken Lang

In [1]:
import pandas as pd
import numpy as np
from sklearn.metrics import confusion_matrix

In [2]:
test = pd.read_csv('test.txt',header=None,sep=' ')
train = pd.read_csv('train.txt',header=None,sep=' ')
dictionary = pd.read_csv('dict.txt',header=None)

In [3]:
train

0    1    2    3    4    5    6    7    8    9    ...  810  811  812  \
0       0    1    3    0    4    0    0    0    6    2  ...    0    0    0   
1       0    0    3    0    1    0    0    1    7    2  ...    0    0    0   
2       0    0    1    0    0    0    0    1    1    2  ...    0    0    0   
3       0    0   12    0    0    0    0    2   12    4  ...    0    0    0   
4       0    1    2    0    0    0    0    2   16    5  ...    0    0    0   
...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
2995    0    0    6    0    2    1    0    2    6    4  ...    0    0    0   
2996    0    0    3    0    3    0    0    0    5    4  ...    0    0    0   
2997    0    0    0    0    1    1    0    0    1    3  ...    0    0    0   
2998    0    0    2    0    0    0    0    0    6    1  ...    0    0    0   
2999    0    0    2    0    1    0    0    0    6    2  ...    0    0    0   

      813  814  815  816  817  818  819  
0       0    0    0    0    0    0    5  
1       0    0    0    0    0    0    1  
2       0    0    0    0    0    0    4  
3       0    0    0    0    0    0    2  
4       0    0    0    0    0    0    2  
...   ...  ...  ...  ...  ...  ...  ...  
2995    0    0    0    0    0    0    4  
2996    0    0    0    0    0    0    1  
2997    0    0    0    0    0    0    6  
2998    0    0    0    0    0    0    1  
2999    0    0    0    0    0    0    5  

[3000 rows x 820 columns]

We wish to predict whether or not a post belongs to group 1 or 2. For the perceptron and logistic regression classifiers, we look at $y:{1,2}$ where $y$ is relabeled ${-1,1}$, respectively.

## Perceptron

We output the label as $y = sign(<w,x>)$. We initialize $w_1 = 0$, and for each pass $t$:

If $y_t<w_t,x_t> \leq 0$: $w_t+1 = w_t + y_t x_t$.
Otherwise, $w_t+1 = w_t$.
    

In [7]:
data = train.where((train.loc[:,819] <= 2)).dropna().reset_index(drop=True)
data

0    1     2    3    4    5    6    7     8    9    ...  810  811  812  \
0     0.0  0.0   3.0  0.0  1.0  0.0  0.0  1.0   7.0  2.0  ...  0.0  0.0  0.0   
1     0.0  0.0  12.0  0.0  0.0  0.0  0.0  2.0  12.0  4.0  ...  0.0  0.0  0.0   
2     0.0  1.0   2.0  0.0  0.0  0.0  0.0  2.0  16.0  5.0  ...  0.0  0.0  0.0   
3     0.0  0.0   5.0  0.0  0.0  0.0  1.0  0.0  12.0  1.0  ...  0.0  0.0  0.0   
4     0.0  0.0   4.0  0.0  0.0  0.0  0.0  0.0   8.0  3.0  ...  0.0  0.0  0.0   
...   ...  ...   ...  ...  ...  ...  ...  ...   ...  ...  ...  ...  ...  ...   
1085  0.0  1.0   1.0  0.0  1.0  0.0  1.0  0.0  17.0  1.0  ...  0.0  0.0  0.0   
1086  0.0  0.0   3.0  0.0  0.0  0.0  0.0  0.0  11.0  3.0  ...  0.0  0.0  0.0   
1087  0.0  0.0   5.0  0.0  1.0  0.0  0.0  0.0  11.0  1.0  ...  0.0  0.0  0.0   
1088  0.0  0.0   3.0  0.0  3.0  0.0  0.0  0.0   5.0  4.0  ...  0.0  0.0  0.0   
1089  0.0  0.0   2.0  0.0  0.0  0.0  0.0  0.0   6.0  1.0  ...  0.0  0.0  0.0   

      813  814  815  816  817  818  819  
0     0.0  0.0  0.0  0.0  0.0  0.0  1.0  
1     0.0  0.0  0.0  0.0  0.0  0.0  2.0  
2     0.0  0.0  0.0  0.0  0.0  0.0  2.0  
3     0.0  0.0  0.0  0.0  0.0  0.0  1.0  
4     0.0  0.0  0.0  0.0  0.0  0.0  2.0  
...   ...  ...  ...  ...  ...  ...  ...  
1085  0.0  0.0  0.0  0.0  0.0  0.0  2.0  
1086  0.0  0.0  0.0  0.0  0.0  0.0  1.0  
1087  0.0  0.0  0.0  0.0  0.0  0.0  2.0  
1088  0.0  0.0  0.0  0.0  0.0  0.0  1.0  
1089  0.0  0.0  0.0  0.0  0.0  0.0  1.0  

[1090 rows x 820 columns]

In [26]:
def perceptron(data, passes):
    x = data.loc[:,:818]
    y = data.loc[:, 819].replace({1:-1,2:1}) #1 is labeled as -1, 2 is 1
    w = np.array([0] * len(x.columns))
    for t in range(passes):
        for feat in range(len(x)):
            if y[feat] * np.dot(w, x.loc[feat]) <= 0: #check projection of y and dot product
                w = w + (y[feat] * x.loc[feat]) #update w if necessary
    return w

In [27]:
def perceptron_err(w, test):
    preds = []
    x = test.loc[:,:818]
    y = test.loc[:, 819]
    for feat in range(len(y)):
        pred = np.dot(w, x.loc[feat]) #2 is >0, 1 is <0
        if pred > 0:
            preds.append(2)
        elif pred < 0:
            preds.append(1)
        elif pred == 0: #dot product is 0, randomly choose
            preds.append(np.random.choice([1,2]))
    return (pd.Series(preds) == y).value_counts(normalize=True)[0]

In [110]:
for x in [2,3,4]:
    print("Pass " + str(x))
    print("Training Error:")
    print(str(perceptron_err(perceptron(data,x), data)) + '\n')
    print("Test Error:")
    print(str(perceptron_err(perceptron(data,x), test.where((test.loc[:,819] <= 2)).dropna().reset_index(drop=True)))+'\n')

Pass 2
Training Error:
0.03853211009174312

Test Error:
0.0610079575596817

Pass 3
Training Error:
0.01926605504587156

Test Error:
0.04509283819628647

Pass 4
Training Error:
0.01651376146788991

Test Error:
0.04509283819628647



## Logistic Regression

We minimize the loss $L(w) = {\Sigma log (1 + e^{−y_iw^{T}x_i})}$ with gradient descent. With $\eta$ = .001, $w_t + 1 = w_t + \eta_t\Sigma \frac{y_ix_i}{1 + e^{y_iw^{T}x_i}}$.

In [20]:
def logistic(data, passes):
    x = data.loc[:,:818]
    y = data.loc[:, 819].replace({1:-1,2:1}) #1 is labeled -1, 2 is 1
    w = np.array([0] * len(x.loc[0]))
    for t in range(passes):
        summ = []
        for i in range(len(x)):
            summ.append((y[i] * x.loc[i])/(1 + np.exp(y[i]*np.dot(w.transpose(),x.loc[i]))))
        w = w + (0.001 * sum(summ)) #learning rate = .001
    return w

In [18]:
def logistic_err(w, test):
    log_pred = []
    x = test.loc[:,:818]
    y = test.loc[:, 819]
    for n in range(len(x)):
        prediction = 1/(1+(np.exp(-(np.dot(w.transpose(),x.loc[n]))))) #1/1+e^-z, sigmoid
        if prediction > .5: #if sigmoid > .5, predict 2, <.5, 1
            log_pred.append(2)
        elif prediction < .5:
            log_pred.append(1)
        elif prediction == .5: #sigmoid is exactly .5, randomly choose
            log_pred.append(np.random.choice([1,2]))
    return (log_pred == y).value_counts(normalize=True)[0]

In [22]:
for x in [10,50,100]:
    print("Pass " + str(x))
    print("Training Error:")
    print(str(logistic_err(logistic(data,x), data)) + '\n')
    print("Test Error:")
    print(str(logistic_err(logistic(data,x), test.where((test.loc[:,819] <= 2)).dropna().reset_index(drop=True)))+'\n')

Pass 10
Training Error:
0.2963302752293578

Test Error:
0.29708222811671087

Pass 50
Training Error:
0.03761467889908257

Test Error:
0.0610079575596817

Pass 100
Training Error:
0.02018348623853211

Test Error:
0.04509283819628647



Let's take a look at the most positive and negative coordinates for both classifiers. Negative and positive coordinates are comp.windows.x and rec.sport.baseball, respectively. Indeed, we see the negative terms are computer related, while the positive terms are sports related.

In [47]:
w3 = perceptron(data,3)

In [42]:
lw3 = logistic(data,50)

### Most Negative

In [48]:
(w3.argsort()[:3]).apply(lambda x:'Coordinate ' + str(x+1) + ': ' + (dictionary.loc[x])[0])

0       Coordinate 439: file 
1    Coordinate 467: program 
2       Coordinate 204: line 
dtype: object

In [49]:
(lw3.argsort()[:3]).apply(lambda x:'Coordinate ' + str(x+1) + ': ' + (dictionary.loc[x])[0])

0    Coordinate 618: window 
1      Coordinate 439: file 
2        Coordinate 73: use 
dtype: object

### Most Positive

In [50]:
(w3.argsort()[-3:]).apply(lambda x:'Coordinate ' + str(x+1) + ': ' + (dictionary.loc[x])[0])

816    Coordinate 394: game 
817    Coordinate 470: team 
818       Coordinate 79: he 
dtype: object

In [51]:
(lw3.argsort()[-3:]).apply(lambda x:'Coordinate ' + str(x+1) + ': ' + (dictionary.loc[x])[0])

816     Coordinate 59: they 
817    Coordinate 394: game 
818       Coordinate 79: he 
dtype: object

## One-vs-all Multi-class Perceptron

What if we want to predict more than just two labels? Let's generate $w$ for every label, where 1 will indicate that the value is that label, and -1 otherwise.

In [7]:
c_i = [0] * len(train[819].unique())
for i in train[819].unique():
    datac = train.copy()
    datac[819] = datac[819].apply(lambda x: 2 if x == i else 1)
    c_i[i-1]= (perceptron(datac, 1))

In [8]:
def ova(c_i, data):
    preds = []
    x = data.loc[:,:818]
    y = data.loc[:,819]
    for i in range(len(c_i)):
        pred_c = []
        for j in range(len(x)):
            pred_c.append(np.dot(x.loc[j],c_i[i]))
        preds.append(pred_c)
    return preds

In [9]:
mat = ova(c_i, test)

In [10]:
ova_df = pd.DataFrame(mat)

Here we can see prediction $w$s for every entry in our test data, among the six labels.

In [11]:
ova_df

0    1     2     3    4     5     6    7    8     9    ...  990  991   992  \
0  -77  122 -2152  -895 -149  -561 -1630  -75 -210  -327  ... -430 -135  -620   
1   36 -281   911  -651  -90  -359  -719  -99  -94  -474  ... -630  101  -790   
2 -266  -73 -2383  -102 -439 -1056   -38  -51 -243   -66  ... -859 -140 -1101   
3 -214  -82 -2561 -1113  -88  -488 -1232   40  288  -536  ... -273    3  -357   
4 -459 -742 -4824 -2172 -309  1758 -1919 -384 -479 -1130  ... -270 -288   901   
5 -107 -661 -1970 -1445  142  -796 -1985 -136 -416   379  ...  502 -297 -1377   

    993  994  995  996  997  998  999  
0  -454 -285   17 -193 -218 -226 -237  
1   413 -306 -100 -151 -264 -291 -559  
2 -1067 -331 -128 -226   59 -121 -861  
3  -346 -241 -112 -242  -88 -234 -499  
4  -891 -244 -415    2 -373  -64 -638  
5  -906 -413 -281 -419  -31  278   22  

[6 rows x 1000 columns]

If there is only one positive value in a column, it means the location of that positive value is predicted to be the corresponding label. Otherwise, the classifier is unsure (e.g. there are no or multiple positive values).

In [14]:
def predict(col):
    know = 0
    loc = 0
    for x in np.arange(len(col)):
        if col[x] > 0:
            loc = x
            know += 1
        elif col[x]==0:
            if np.random.choice([0,1]) == 0:
                know += 1
    if know == 1:
        return loc+1
    return 0
    

In [15]:
confusion = pd.DataFrame(confusion_matrix(test.loc[:,819], ova_df.apply(predict),normalize='true'))

In [27]:
confusion.drop(0,axis=0).transpose().rename({0:"Unsure"})

1         2         3         4         5         6
Unsure  0.232432  0.270833  0.451429  0.255435  0.115385  0.333333
1       0.718919  0.010417  0.034286  0.021739  0.000000  0.009259
2       0.010811  0.656250  0.034286  0.027174  0.012821  0.018519
3       0.000000  0.015625  0.371429  0.000000  0.000000  0.027778
4       0.016216  0.005208  0.000000  0.690217  0.000000  0.000000
5       0.016216  0.031250  0.074286  0.005435  0.801282  0.120370
6       0.005405  0.010417  0.034286  0.000000  0.070513  0.490741

In [26]:
(test.loc[:,819] == ova_df.apply(predict)).value_counts(normalize=True)[1]

0.629

Our one-vs-all classifier is 63% accurate overall on the test data. It does particularly well at predicting posts from talk.politics.mideast, and performs worst on sci.med.